### PR TITLE
storage: Consider which replica will be removed when adding a replica…

### LINKED ISSUE #17971: storage: Consider which replica will be removed when adding a replica to improve balance
In clusters with multiple localities, we try to maintain diversity by spreading the replicas for each range as evenly across localities as possible. This means that in a cluster with 3 localities and 3 replicas per range, we'll try to keep ` replica in each locality for each range.

When in a rebalancing state where one locality has 1 replicas and the other localities each have 1, we'll always remove a replica from the locality that has 2. This is good.

What isn't as good is that we don't consider this when deciding whether to rebalance in the first place. Our rebalancing logic will kick in if a possible new destination is a better fit than any of the existing replicas, even if the new destination is in a different locality and thus won't actually be a direct replacement for the existing replica that isn't a great fit.

I haven't seen this cause massive problems by itself, but in combination with another problem (I've seen it flare up with both #17879 and #17970) it can make for rebalance thrashing, where we repeatedly add and remove a replica on the same 1 or 2 nodes.

Not all situations will be as straightforward as the 3-locality example above, so we'll have to make sure the fix for this is somewhat more general than just making sure that a potential replica to add is in the same locality as the worst existing replica.

### PR DESCRIPTION
**Pull request by a6802739**
@a-robinson 

Fixes #17971. @a-robinson , Thanks for your guidance, I just simulate `(*Allocator).RemoveTarget` before we actually add the replica during the rebalance.

Still don't know how to add a test for this.

### CODE REVIEW

**Comment by a-robinson** [comment]
Hmm, I think we may need to take this a little farther. `RebalanceTarget` might internally find multiple good stores to rebalance a range to. It only returns the one that it thinks is best, but it may have found others that would be really good fits. In such cases, we should rebalance to one of the other candidates, but this code will just return without doing a rebalance. I think we need to do this sort of filtering inside `RebalanceTarget` so that it can return a different candidate if there is one that would work instead.

As for testing, the replicate queue is unfortunately in need of some test love (#11987). If you could write one that covers this code, that would be great, otherwise some manual testing may be needed. The best tests for it currently are the allocator acceptance tests and running the `cmd/allocsim` tool and inspecting the output by hand.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replicate_queue.go, line 498 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364#-Ktso9nkcPen3-l0uuLK:-Ktso9nl_Dq7DXk0EApI:b8pwnk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/321f8b26853abac4528504007b20eccff2b77b59/pkg/storage/replicate_queue.go#L498)):*
> ```Go
> 					StoreID: rebalanceReplica.StoreID,
> 				}
> 				candidates := filterUnremovableReplicas(repl.RaftStatus(), desc.Replicas, 0)
> ```

I think we'd be best off if we simulate the removal process as closely as possible, which would mean including the new replica in the `RaftStatus` with a `Progress` of 0 and passing an ID for it as the `brandNewReplicaID` to `filterUnremovableReplicas` rather than appending it to the list of candidates afterwards.

---

*[pkg/storage/replicate_queue.go, line 500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364#-KtsotNJ-2N2m-lKVSVU:-KtsotNKWOMxIjoizm9j:b-8etsfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/321f8b26853abac4528504007b20eccff2b77b59/pkg/storage/replicate_queue.go#L500)):*
> ```Go
> 				candidates := filterUnremovableReplicas(repl.RaftStatus(), desc.Replicas, 0)
> 				candidates = append(candidates, rep)
> 				removeReplica, _, err := rq.allocator.RemoveTarget(ctx, zone.Constraints, candidates, rangeInfo)
> ```

I'm a little concerned about how this will interact with #18425, since we aren't using the updated stats here. Do you have any thoughts on how to make this more accurately reflect the updated stats? I think we may be best off not doing anything about it, but I imagine this discrepancy will probably be a problem at some point.

---

*[pkg/storage/replicate_queue.go, line 505 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364#-KtspxPE6PMeWDSpIhK5:-KtspxPE6PMeWDSpIhK6:biw5jg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/321f8b26853abac4528504007b20eccff2b77b59/pkg/storage/replicate_queue.go#L505)):*
> ```Go
> 				}
> 				if removeReplica.StoreID == rebalanceReplica.StoreID {
> 					log.VEventf(ctx, 1, "we may immediately remove the same replica after we add this replica for rebalancing")
> ```

If you don't mind me rewording this, I'd phrase it as "not rebalancing to candidate %+v because it would be removed immediately after being added"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364#-:-KtsqSKznU6TP0MIqQEl:bcycgon)*

**Comment by a6802739** [comment]
@a-robinson , Sorry, I couldn't login in Reviewable right Now.

If we already pass the new replica ID as `brandNewReplicaID` to `filterUnremovableReplicas`, why we still include the new replica in the RaftStatus with a Progress of 0?


And what do you mean `I'm a little concerned about how this will interact with #18425, since we aren't using the updated stats here`, I'm sorry, I don't quite understand here? Do you mean we should add the update stats first when we try to simulate the remove procedure?

**Comment by a-robinson** [comment]
Sorry, I'll try to stick to Github comments if Reviewable is inconvenient for you.

> If we already pass the new replica ID as `brandNewReplicaID` to `filterUnremovableReplicas`, why we still include the new replica in the RaftStatus with a Progress of 0?

Because it makes a difference in what `filterUnremovableReplicas` will do. If the new replica isn't included in the `RaftStatus` then `filterUnremovableReplicas` will behave as if it doesn't exist (even if we pass it in as the `brandNewReplicaID`). If we add the new replica to the `RaftStatus` but don't pass it in as the `brandNewReplicaID`, then `filterUnremovableReplicas` will consider the new replica to be behind, which will make it likely to be selected.

> And what do you mean `I'm a little concerned about how this will interact with #18425, since we aren't using the updated stats here`, I'm sorry, I don't quite understand here? Do you mean we should add the update stats first when we try to simulate the remove procedure?

The goal here is to simulate the actual removal process as accurately as possible. If the actual removal process will be acting on updated stats and this simulated removal process isn't, then the simulated process may be incorrect sometimes. Modifying the stats for the simulation in a way that's easy to roll back and doesn't affect anything else happening in the system may be difficult, though. What do you think?

**Comment by a6802739** [comment]
> The goal here is to simulate the actual removal process as accurately as possible. If the actual removal process will be acting on updated stats and this simulated removal process isn't, then the simulated process may be incorrect sometimes. Modifying the stats for the simulation in a way that's easy to roll back and doesn't affect anything else happening in the system may be difficult, though. What do you think?

I think the actual intent is just to simulate `RemoveTarget`, but we didn't actually did the rebalance before the simulation. If we want to be very close to the real `RemoveTarget`, I think we we just to to update the local copy of target rebalance store, I mean  add `WritesPerSecond` for the target store, and recover it after the simulation?

**Comment by a-robinson** [comment]
That would be ideal, yes. It'll probably require a fair amount of refactoring, so use your judgment, but that would be best.

**Comment by a6802739** [comment]
@a-robinson, Thank you for your review. 
I have made a simulation in `RemoveTarget`. Still not sure how to add a test for it.

File: pkg/storage/allocator.go
Comment (a-robinson) [inline pkg/storage/allocator.go]: Do you think we could avoid this code duplication with what's in `replicate_queue.go` and instead create a shared function they can both use? If they're separate like this, it seems likely that they'll accidentally diverge in the future.

File: pkg/storage/allocator.go
Comment (a-robinson) [inline pkg/storage/allocator.go]: I think we actually want this logic to happen in a loop. What if the `newTarget` is also going to be immediately removed?

**Review by a-robinson — COMMENTED** [review]
I'm sorry for forgetting about this for a while, @a6802739! We definitely still want this to get in.

Regarding testing, we have a few options of increasing difficulty:

1. Do something like in #18425, where we create a configuration that would previously have caused thrashing and test that `RebalanceTarget` will prefer not rebalancing over rebalancing to something that will immediately be removed.
1. Create a test cluster with multiple nodes via something like `multiTestContext` or `TestCluster` with localities configured in a way that would normally trigger this bug and ensure that the rebalancing settles down. For an example configuration, see #19013. To summarize that issue, a cluster with 2 nodes in locality X, 1 node in locality Y, and 1 node in locality Z will exhibit the kind of thrashing that this PR should fix.
1. Most complicated would be to randomly generate cluster/replica configurations to run through a unit test that tries runs `RebalanceTarget`, makes (or simulates) the requested change, and then runs `RemoveTarget`. The test would fail if any configuration chose to remove the replica that it just added.

**Comment by a-robinson** [comment]
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364)*

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: Hmm, Sorry, I couldn't quite understand what do you mean here. I didn't make any change to `replicate_queue.go`.
Would you please point out the  code duplication with `replicate_queue.go`? Thank you very much.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: Great. Done.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: My mistake. Done.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: Sure, Done.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: Yeah, You are right. My mistake, Done.

**Comment by a6802739** [comment]
@a-robinson, Thank you for your reference about unit test. Please have a review again. Thank you very much.

**Comment by a6802739** [comment]
@a-robinson , Sorry, I don't know why  the teamcity failed. And I didn't see any important information from the teamcity.

File: pkg/storage/allocator.go
Comment (a-robinson) [inline pkg/storage/allocator.go]: Passing in a `Replica` is pretty broad. It looks like we could shrink this down to just using the `RangeInfo` if we made `simulateRemoveTarget` just take a `RangeInfo` instead of a `Replica`.

File: pkg/storage/allocator_test.go
Comment (a-robinson) [inline pkg/storage/allocator_test.go]: Can you confirm that this test passes with your change but fails without it?

File: pkg/storage/allocator_test.go
Comment (a-robinson) [inline pkg/storage/allocator_test.go]: Can you add a comment explaining these stores and the fake replica stats below? I'm not really sure why this fifth store is here or why are the values are set the way that they are. And if I'm not sure, then future readers are even less likely to understand what you were thinking while creating the test.

File: pkg/storage/allocator_test.go
Comment (a-robinson) [inline pkg/storage/allocator_test.go]: Why? In 1.1, this problem occurs even if stats-based rebalancing is disabled. Testing the default setting seems more valuable than testing the non-default setting. Keeping this false would also let us remove the capacity, available, logicalbytes, and writespersecond fields from all the store descriptors above.

File: pkg/storage/allocator_test.go
Comment (a-robinson) [inline pkg/storage/allocator_test.go]: I don't think having a fifth store here is any better than just having 4 stores and verifying that we don't want to move a replica to the fourth.

**Review by a-robinson — COMMENTED** [review]
I think that test failure in teamcity is just a flake - https://github.com/cockroachdb/cockroach/issues/18554

File: pkg/storage/allocator.go
Comment (a-robinson) [reply pkg/storage/allocator.go]: I was referring to the code for the `AllocatorRemove` case in `processOneChange`, but on looking at it again it doesn't look like there's much that could be de-duped. Sorry for the confusion!

**Comment by a-robinson** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364)*

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: Yes, I could confirm that.

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: Hmm, my original thinking is that if we didn't add the fifth store, it will have no store to make an rebalance, So I just add a fifth store here. But you are right, it seems it have no need to add the fifth store here, it could just return a nil target for `RebalanceTarget`.

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: Yeah, you are right.

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: hmm, Yeah, that will be more easier. Thank you very much.

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: Yeah, you are right.  4 stores will be enough.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: So we should guarantee the duration for `WritesPerSecond` in `RangeInfo` exceeds `MinStatsDuration`?

Comment (a-robinson) [reply pkg/storage/allocator.go]: Yes, I suppose so. Since this code only runs on the leaseholder, it pretty much always should be, anyway.

File: pkg/storage/allocator.go
Comment (a-robinson) [inline pkg/storage/allocator.go]: Sorry for not noticing this last time, but `for len(candidates) >= 0` is a pretty meaningless loop condition. Did you mean `for len(candidates) > 0`? Or to just make this an infinite loop?

File: pkg/storage/allocator_test.go
Comment (a-robinson) [inline pkg/storage/allocator_test.go]: Thanks for explaining in the code review, but a comment in the code explaining this configuration of stores would be very helpful to future maintainers.

**Review by a-robinson — COMMENTED** [review]
Just a couple small comments left. Should be good to go after the last couple changes :)

**Comment by a-robinson** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364)*

File: pkg/storage/allocator_test.go
Comment (a6802739) [reply pkg/storage/allocator_test.go]: Sorry for that. I have already added a comment about the configuration of stores.

File: pkg/storage/allocator.go
Comment (a6802739) [reply pkg/storage/allocator.go]: Yeah, I just thought it will break in the loop condition. I have changed it as your suggestion.

**Review by a-robinson — APPROVED** [review]
LGTM, thanks!

**Comment by a-robinson** [comment]
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18364)*

**Review by a-robinson — COMMENTED** [review]
It looks like it fails to build, though, according to teamcity:

```
[TestStyle/TestErrCheck] style_test.go:819: err=exit status 2, stderr=/go/src/github.com/cockroachdb/cockroach/pkg/storage/store_pool_test.go:434:45: cannot use replica (variable of type *Replica) as RangeInfo value in argument to sp.updateLocalStoreAfterRebalance
```

**Comment by a6802739** [comment]
@a-robinson, Thank you very much. I have changed it already. Wait for a successful CI of the teamcity. :)

**Pull request merged**
